Ignore SyntaxError while parsing

c-w opened the pull request:
This is a partial fix for #168. With this change, the parser no longer crashes on test file [E90.py](https://github.com/PyCQA/pep8/blob/fed43c5/testsuite/E90.py).

Note that this is a pretty minimal fix. A potential future refactor to increase the robustness of the parser could be to use the null-object pattern and return a NullToken from `TokenStream.move` instead of `None`. In this way, any code that consumes `TokenStream.current` doesn't have to null-check the property value explicitly which should make the other issues mentioned in #168 go away.

sigmavirus24 commented:
Could you add tests for this, please?

c-w commented:
@sigmavirus24: Sure thing. 9615c72 adds a test for the fix in dc34871. This also required making a fix in the test harness: 839b528.

Nurdok commented on src/pydocstyle/parser.py:
```diff
@@ -210,14 +210,22 @@ def __init__(self, filelike):
+        try:
+            return next(self._generator, None)
+        except SyntaxError as syntax_error:
+            self.log.warning('error generating tokens: {0}', syntax_error)
```
Instead of capturing the exception and passing it, you can do:
```.py
self.log.warning('error generating tokens', exc_info=True)
```

c-w replied:
Good shout.

Nurdok commented on src/tests/test_definitions.py:
```diff
@@ -126,6 +126,36 @@ def nested_3(self):
+toto = 0 + 0
+#:
+"""  # noqa
+
```
Why do you need this huge snippet? You can create a syntax error with a single character (like "[").

c-w replied:
This snippet is the test file that was reported to cause issues in #168, but yes, we can minimize the test example. Turns out that the single invalid token (e.g. "[" like you mentioned) leads to a different exception: tk.TokenError so I added that as a test case too.

c-w commented:
@Nurdok: Addressed comments and resolved merge conflicts.

Nurdok commented on src/tests/test_definitions.py:
```diff
@@ -126,6 +126,16 @@ def nested_3(self):
+    pass
+"""
+
+source_token_error = """
```
Why do you need this to be a multi-line string? Why not just `source_token_error = "["`?

c-w replied:
Done.

Nurdok commented on src/pydocstyle/parser.py:
```diff
@@ -209,14 +209,22 @@ def __init__(self, filelike):
+    def _next_from_generator(self):
+        try:
+            return next(self._generator, None)
+        except (SyntaxError, tk.TokenError):
```
Please also add this bugfix to the release notes.

c-w replied:
Added a line to `docs/release_notes.rst`.

c-w commented:
@Nurdok: Addressed the latest set of comments, added the fix to the release notes and did some commit squashing because the pull request was getting rather long.

Nurdok commented on docs/release_notes.rst:
```diff
@@ -11,6 +11,10 @@ Major Updates
 
+Bug Fixes
+
+* Made parser more robust to bad source files (#168)
```
Include the pull request as well. i.e., this should be:

    * Made parser more robust to bad source files (#168, #214).

c-w replied:
Thanks. Added.

Nurdok commented:
Merged. Thanks!
